[28기 안현재][Add] Misson7 제외 기능 완성, 리팩토링 중

hyeonze opened the pull request:
- Login 및 Main  페이지 레이아웃 완성
- 로그인 및 댓글기능 완성

## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- main페이지 댓글개시 기능 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 댓글개시 기능 구현
- 개시버튼을 누르거나, 댓글창에서 Return키를 입력하면 addComment함수 실행됨
- addComment함수는 댓글창 내용을 가져와 댓글내용에 추가함
- 테스트 방법 : 댓글 내용을 입력하고, 개시버튼 또는 Return키 입력

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- keyup이벤트 사용시 body전체에서 addEventListenr사용했는데 바꾸는게 좋을 것 같습니다.
- 댓글이 여러개 달렸을때 로직을 구현하지 못했습니다.

<br />

## :: 기타 질문 및 특이 사항
- 함수 로직이 깔끔하지 못한 것 같아 리팩토링 하고있습니다.
- 개인적인 고민과 동료들 의견 반영해서 리팩토링 했으나 오히려 가독성은 더 떨어지는 것 같습니다.

yojuyoon commented on students/14th/eommoonjoo/js/login.js:
```diff
@@ -0,0 +1,31 @@
+
+let input= document.querySelector("input")
+
+id.addEventListener('keyup', function() {
+    if (id.value == null || id.value == "" ) {
+        button.style.backgroundColor = "black";
+    } else {
+        button.style.backgroundColor = "blue";
+    }
+
+pw.addEventListener('keyup', function() {
+    if (pw.value == null || pw.value == "" ) {
+        button.style.backgroundColor = "black";
+    } else {
+        button.style.backgroundColor = "blue";
+    }
```
- 질문 주신 내용이네요 문주님! 동일 이벤트를 따로 적용시켜야 하는 이유는 event.target인 ID와 PW가 각각의 객체타입으로 addEventListener 메서드를 사용하기 때문입니다. 때문에 동일한 이벤트를 함께 적용시키고자 하신다면 최상위 부모인 `<form>`태그에서 이벤트 캡쳐링을 적용해주면 가능합니다.
- 좀 더 공부해보시려면 event target, event bubbling, event capture 를 검색해보세요 :)

eommoonjoo replied:
감사합니다. 지영님 ^^ 말씀해 주신 부분을 생각해서 적용을 해보니, 작동을 하는 것을 확인 하였습니다. 추가적으로 더 알아보니, 3항 연산자라는 것이 있는 것을 알게 되었습니다. 제가 만든 3개의 함수를 한 줄로 해결 할 수 있는 방법이 있더라구요... 
적용하니, 제가 원하는대로 작동을 했습니다.

eommoonjoo replied:
```suggestion
id.addEventListener('keyup', function() {
    if (id.value == null || id.value == "" ) {
        button.style.backgroundColor = "black";
    } else {
        button.style.backgroundColor = "blue";
    }

pw.addEventListener('keyup', function() {
    if (pw.value == null || pw.value == "" ) {
        button.style.backgroundColor = "black";
    } else {
        button.style.backgroundColor = "blue";
    }
```
```
document.addEventListener('keyup', fuction() {
  button.style.backgroundColor = id.value && pw.value ? "green" : black"});
```

yojuyoon commented on students/14th/eommoonjoo/js/login.js:
```diff
@@ -0,0 +1,31 @@
+
+id.addEventListener('keyup', function() {
+    if (id.value == null || id.value == "" ) {
+        button.style.backgroundColor = "black";
+    } else {
+        button.style.backgroundColor = "blue";
```
스타일 속성을 적용할 때에는 css에서 적용해주시고, 해당 속성이 적용된 class이름을 적용시켜주는 방법이 있습니다.
- classList.add 를 사용해주는 것인데요 관련하여 더욱 살펴보시려면 공식 문서를 살펴보세요! 
- https://developer.mozilla.org/ko/docs/Web/API/Element/classList

yojuyoon commented on students/14th/eommoonjoo/js/login.js:
```diff
@@ -0,0 +1,31 @@
+document.getElementById('logInBtn').onclick = function() {
+    let logInId = document.getElementById('logInId').value;
+    let logInPw = document.getElementById('logInPw').value;
+    if ( logInId === "a" && logInPw === "a") {
+        alert("로그인 성공!");
+    } else {
+        alert("로그인 실패!");
+    }
```
```suggestion
  if (logInId === "a" && logInPw === "a") return alert("로그인 성공!");
  alert("로그인 실패!");
```
- 이런식으로 if문에서 미리 return 해주시면 else문을 쓰지 않아도 됩니다.
- if-else분기문을 지양하도록 권장드리고있는데, 우선은 사용하시면서 리뷰 반영을 통해 차차 익혀주시면 좋을 것 같습니다 :)

eommoonjoo replied:
- else를 안쓸 수 있어서 행복해요~~^^

yojuyoon commented on students/14th/eommoonjoo/login.html:
```diff
@@ -0,0 +1,40 @@
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width">
+    <title>repl.it</title>
```
사이트에 맞게 title도 센스있게 바꿔주시는게 어떨까요? 😊

yojuyoon commented on students/14th/eommoonjoo/login.html:
```diff
@@ -0,0 +1,40 @@
+    
+    <link href="style/common.css" rel="stylesheet" type="text/css" />
+     <link href="style/login.css" rel="stylesheet" type="text/css" />
+      
+  
```
불필요한 엔터는 지워주셔도 좋을 것 같습니다 :)

yojuyoon commented on students/14th/eommoonjoo/login.html:
```diff
@@ -0,0 +1,40 @@
+  <body>
+    <div id="wrapper">
+      
+      <div class="main">
```
`<main>`태그를 활용해볼 수도 있겠네요!

yojuyoon commented on students/14th/eommoonjoo/login.html:
```diff
@@ -0,0 +1,40 @@
+            <input type="text" id="logInId" placeholder="전화번호, 사용자 이름 또는 이메일"><br>
+            <input type="password" id="logInPw"  placeholder="비밀번호"><br>
+            <button type="button" id="logInBtn">로그인</button><br>
+            <div class ="ddo">또는</div><br>
```
- 또는이라 ddo 일까요,,?😱
- div는 block 요소입니다! 혹시 `<br>`태그를 사용해주신 이유가 있으실까요?

yojuyoon commented on students/14th/eommoonjoo/login.html:
```diff
@@ -0,0 +1,40 @@
+            <input type="password" id="logInPw"  placeholder="비밀번호"><br>
+            <button type="button" id="logInBtn">로그인</button><br>
+            <div class ="ddo">또는</div><br>
+            <div class="hyper">
```
hyper라는 클래스이름이 조금 모호한 것 같습니다. facebookLink 와 같이 좀 더 명확한 이름을 지어주시면 좋을 것 같습니다 :)

yojuyoon commented on students/14th/eommoonjoo/main.html:
```diff
@@ -0,0 +1,25 @@
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width">
+    <title>repl.it</title>
```
여기도 바꿔주실 수 있겠네요 :)

yojuyoon commented on students/14th/eommoonjoo/main.html:
```diff
@@ -0,0 +1,25 @@
+        <li><a href = "instagram.com"><img src = "https://s3.ap-northeast-2.amazonaws.com/cdn.wecode.co.kr/bearu/explore.png"></a></li>
+        <li><a href = "instagram.com"><img src = "https://s3.ap-northeast-2.amazonaws.com/cdn.wecode.co.kr/bearu/heart.png"></a></li>
+        <li><a href = "instagram.com"><img src = "https://s3.ap-northeast-2.amazonaws.com/cdn.wecode.co.kr/bearu/profile.png"></a></li>
+
+
```
불필요한 엔터는 삭제해주세요! 🙂

yojuyoon commented on students/14th/eommoonjoo/style/login.css:
```diff
@@ -0,0 +1,103 @@
+* { box-sizing: border-box;}
```
```suggestion
* {
 box-sizing: border-box;
 }
```
- 속성은 이런식으로 작성해주시면 좋을 듯 합니다 :)

yojuyoon commented on students/14th/eommoonjoo/style/login.css:
```diff
@@ -0,0 +1,103 @@
+
+#wrapper {
+    width: 380px;
+    margin-left:auto;
+    margin-right:auto;
```
두 속성을 합쳐줄 수 있을 것 같네요!

yojuyoon commented on students/14th/eommoonjoo/style/login.css:
```diff
@@ -0,0 +1,103 @@
+    margin-left:auto;
+    margin-right:auto;
+    height: 600px;
+}
+
+
+
+
+.main {
```
새로운 selector사이는 한줄만 띄워주시면 좋을 것 같습니다. 너무 붙어있어도, 너무 떨어져있어도 가독성에 어려움을 줄 수 있기 때문입니다 :)

yojuyoon commented on students/14th/eommoonjoo/style/login.css:
```diff
@@ -0,0 +1,103 @@
+
+.main {
+    border: 1px solid;
+    margin-top: 100px;
+    margin-left: auto;
+    margin-right: auto; 
```
이 속성 또한 하나로 합쳐주실 수 있을 것 같네요!

yojuyoon commented on students/14th/eommoonjoo/style/login.css:
```diff
@@ -0,0 +1,103 @@
+    margin-top: 100px;
+    margin-left: auto;
+    margin-right: auto; 
+    padding-bottom: 50px;
+    padding-top: 20px;
```
마찬가지입니다 :)

yojuyoon reviewed (commented):
문주님! 차근차근 진행중이시군요!🔥 우선 질문 주신 내용에 대한 구체적인 이야기는 답변으로 달아드렸습니다!
두가지 추가적으로 말씀드릴 부분은 함수 작성을 하는 여러가지 방법이 있는데요, 우선 DOM에 접근하여 변수를 할당한 뒤 실행하고자 하는 함수를 따로 작성해주는 방법도 있습니다! 

현재는 해당 DOM에 바로 addEventListener로 적용시켜주고 계시지만, 해당 함수를 따로 생성하여 addEventListener에 함수명만 작성해주는 방법이 있습니다! 
관련된 공식문서 링크로 달아드립니다. 함께 읽어보며 공부해보세요 :)

또 클래스명에 대해서도 조금 더 고민하여 지어주시는게 좋을 것 같습니다! 원래 클래스명 짓는게 어렵습니다..😂 문주님 화이팅! 
- https://developer.mozilla.org/ko/docs/Web/API/EventTarget/addEventListener

Commit:
[Add] Login 및 Main 레이아웃 && 로그인 기능과 댓글기능 완성
- Login 및 Main  페이지 레이아웃 완성
- 로그인 및 댓글기능 완성

Commit:
Refactor: 스스로 리팩토링
- login.js : body태그 전체에 addEventListener메서드 사용했었는데 input태그에서만 addEventListener메서드 사용했을 때로 변경
- main.js : login.js와 마찬가지로 keyup이벤트의 변수 수정

Commit:
Add: main.html 댓글개시 후 개시내용 제거
- 댓글개시 후에도 input 태그에 내용이 남아있는 문제 해결

Commit:
Add : 댓글기능 개별 좋아요버튼
- 댓글 추가시 각자 좋아요버튼이 생기도록 수정했습니다

Commit:
Add : 댓글 달 경우 화면높이가 늘어나는 기능 구현
- 댓글을 많이달면 다 표시하지 못하는 문제 해결했습니다.
- 왼쪽 요소에 height값이 있어 내용이 늘어나도 높이가 유지됐었는데
- height값 제거 한 후 댓글창은 아래로 배치했습니다.

Commit:
Add: 이전 커밋 후 배경색상 제거
- layout 수정 위해 임시추가했던 background-color 제거했습니다.

Commit:
Add: 엔터키로 로그인 기능 구현
- 로그인버튼으로만 로그인 가능했었는데
- 엔터키로도 가능하도록 했습니다.

Commit:
Add: Misson5 | ID, PW validation
아이디 / 비밀번호 유효성검사 기능 추가했습니다.
아이디는 @가 포함되어야 하고 비밀번호는 5자이상일 때
로그인버튼 활성화됩니다.

Commit:
Add: 댓글 좋아요 기능 및 삭제기능 추가
- 댓글란 끝쪽에 하트를 누르면 눌러집니다.
- 댓글을 추가한 후 x버튼을 누르면 삭제됩니다.

Commit:
Add: 댓글 제거버튼 호버했을 때만 나타나게 하는 게능 추가
- css :hover로 구현했습니다.
- 수정 후 커밋이 안돼서 공백추가후 재시도

Commit:
Add: 프로필메뉴 토글기능 && 스토리창 이미지 보더 적용
- 프로필을 누르면 메뉴가 나타나는 기능 추가했습니다.(Mission 8)
- 스토리가 있는 유저의 프로필에 css로 보더를  추가했습니다.

Commit:
Add: nav에 z-index 3 줘서 스토리가 올라가는 현상 해결

Commit:
Add: placehoder에 돋보기 백그라운드로 추가하여 글자입력시 돋보기도 사라지는 기능 구현

Commit:
Fix : 댓글 추가 후에도  게시버튼이 활성화돼있는 상태
- 댓글 추가 후에 게시버튼 활성화되는 문제를 해결했습니다.
- addComment(댓글 추가하는 함수) 끝에 댓글 추가 후 게시버튼의
- 상태를 원복시키는 코드 추가했습니다.

Commit:
Refactor: 콜백함수 분리하여 가독성 향상
- addEventListener함수의 인자인 콜백함수들을 분리하여 작성했습니다

Commit:
Fix: main.js Line 50 조건문 한 줄로 수정

Commit:
Fix: main.html 주석제거

Commit:
Add : 검색기능 구현
- main.html에서 검색창의 검색기능 구현
- 아직 코드 이해 및 리팩토링이 부족합니다.

Commit:
Add: 검색기능 이해 완료 &&  주석 추가
- 구현한 검색기능을 완벽히 이해했고 이해한내용을 주석으로 남겼습니다.
- 리팩토링이 필요합니다.

Commit:
Refactoring : searchInput의 콜백함수 분리

Commit:
코드정리 후 커밋합니다

Commit:
Fix : 검색결과 미디어쿼리 대응
상단 검색바에서 검색결과가 초기상태에서 안보여야 하는데
미디어쿼리 1000px이하에서 대응하지 못하는 에러 수정

Commit:
Fix : 로그인페이지 하단 a태그 위치수정
